packages/shared/src/contexts/ActiveFeedContext.tsx

```diff
@@ -1,11 +1,118 @@
-import { createContext } from 'react';
+import React, {
+  ReactElement,
+  ReactNode,
+  useContext,
+  useMemo,
+  useRef,
+} from 'react';
 import { FeedReturnType } from '../hooks/useFeed';
+import { Post } from '../graphql/posts';
+import useLeanPostActions from '../hooks/post/useLeanPostActions';
 
 export type ActiveFeedContextValue = {
   queryKey?: unknown[];
+  pinPost?: boolean;
   items: FeedReturnType['items'];
+  onClick?: (post: Post) => void;
+  onBookmark?: (post: Post) => void;
+  onUpvote?: (post: Post) => void;
+  onDownvote?: (post: Post) => void;
+  onHidePost?: (post: Post) => void;
+  onPromotePost?: (post: Post) => void;
+  onBanPost?: (post: Post) => void;
+  onDeletePost?: (post: Post) => void;
+  canDeletePost?: (post: Post) => boolean;
+  canPinPost?: (post: Post) => boolean;
+  onPinPost?: (post: Post) => void;
+  onBlockSource?: (post: Post) => void;
+  onBlockTag?: (post: Post, tag: string) => void;
+  feedRef?: React.RefObject<HTMLDivElement>;
 };
 
-export const ActiveFeedContext = createContext<ActiveFeedContextValue>({
+const ActiveFeedContext = React.createContext<ActiveFeedContextValue>({
   items: [],
 });
+
+export default ActiveFeedContext;
+
+type ActiveFeedContextProviderProps = ActiveFeedContextValue & {
+  children: ReactNode;
+};
+
+export const ActiveFeedContextProvider = ({
+  children,
+  items,
+  queryKey,
+  onClick,
+  pinPost = true,
+}: ActiveFeedContextProviderProps): ReactElement => {
+  const feedRef = useRef();
+  const virtualGrid =
+    feedRef.current && getComputedStyle(feedRef.current).gridTemplateColumns;
+
+  // TODO: determine tracking when lean post actions are invoked from a feed type
+  const {
+    onPromotePost,
+    onBanPost,
+    canDeletePost,
+    onDeletePost,
+    canPinPost,
+    onPinPost,
+    onBlockSource,
+    onBlockTag,
+    onClick: onClickDefault,
+    onBookmark,
+    onUpvote,
+    onDownvote,
+    onHidePost,
+  } = useLeanPostActions({
+    queryKey,
+  });
+
+  // console.log(virtualGrid);
+  const data: ActiveFeedContextValue = useMemo(
+    () => ({
+      items,
+      queryKey,
+      onClick: onClick || onClickDefault,
+      onBookmark,
+      onUpvote,
+      onDownvote,
+      onHidePost,
+      onPromotePost,
+      onBanPost,
+      canDeletePost,
+      onDeletePost,
+      canPinPost: pinPost && canPinPost,
+      onPinPost,
+      onBlockSource,
+      onBlockTag,
+      feedRef,
+    }),
+    [
+      items,
+      onClick,
+      onBookmark,
+      onUpvote,
+      onDownvote,
+      onHidePost,
+      onPromotePost,
+      onBanPost,
+      canDeletePost,
+      onDeletePost,
+      canPinPost,
+      onPinPost,
+      onBlockSource,
+      onBlockTag,
+      queryKey,
+    ],
```

Comment:
We need to be careful, this currently is going to make feed rendering as slow as it is now or worse. 

Basically it should function the same as `LeanOptionsButton` where hooks down the render tree call `useLeanPostActions` and context only provides bare minimum like items and queryKey.

Reply:
You do the custom logic in that specialized component?

Reply:
But let's say it's `canPinPost` the page itself defines this.
The specialized component would have no way of knowing about it, unless it's contexted/prop drilled correct? (as it's implementation sits quite deep)

Reply:
`canPinPost` would be something that is in the context then yes and component can decide what to do with it.